lib/fs.js

```diff
@@ -2337,6 +2337,71 @@ function writeFileSync(path, data, options) {
   }
 }
 
+/**
+ * walk and traverse the file directory tree.
+ * @param {string | Buffer | URL} path
+ * @param {{
+ *    dir?:array;
+ *    file?:array;
+ * }} [options]
+* @param {(
+*     root?:string,
+*     dirs?:array,
+*     files?:array
+*  ) => any} callback
+* @returns {void}
+*/
+function walk(path, options={ dir:[], file:[] }, callback) {
+  path = getValidatedPath(path);
+  callback = maybeCallback(callback || options);
+  if(!options.dir){
+    options.dir=[];
+  }
+  if(!options.file){
+    options.file=[];
+  }
+  
+  var root="";
+  for(var j = 0;j < path.length ;j++){
+    if(path[j] == "\\"){
+      root += "/";
+    }else{
+      root += path[j];
+    }
+  }//var root = top.replaceAll("\\","/");
+  if(root[root.length-1] == "/"){
+    root = StringPrototypeSlice(root,0,-1);// root.slice(0, -1);
+  }
+
+  function walk_in(arr, item){
+    for(var i = 0 ;i < arr.length ;i++){
+      if(arr[i] == item){
+        return false;
+      }
+    }
+    return true;
+  }
+  function walk_check(root){
+    var dirs = [], files = [];
+    let filesArr = readdirSync(root, { encoding: 'utf8' });
+    for(var j = 0 ;j < filesArr.length ;j++){
+      var item=filesArr[j];
+      if(statSync(root + "/" + item).isDirectory() && walk_in(options.dir,item)){
+        ArrayPrototypePush(dirs, root + "/" + item);//dirs.push(root + "/" + item);
+      }else if(statSync(root + "/" + item).isFile() && walk_in(options.file,item)){
+        ArrayPrototypePush(files,item);//files.push(item);
+      }
+    }
+    callback(root,dirs,files);
```

Comment:
This is going to call the callback multiple times. That's not consistent with the Node.js API design. This could work as an event emitter perhaps, but not a callback.